src/daemon/delayed_shutdown_timer.cpp

```diff
@@ -16,78 +16,87 @@
-        {
-            ssh_session->exec(fmt::format("wall The system is going down for poweroff now"));
-        }
+        vm.ssh_exec(cmd);
+    }
+    catch (const mp::SSHException& e)
+    {
+        mpl::log(mpl::Level::info, vm.vm_name, fmt::format("Could not broadcast shutdown message in VM: {}", e.what()));
+    }
+}
+
+void write_shutdown_message(mp::VirtualMachine& vm, const std::chrono::milliseconds& time_left)
+{
+    if (time_left > std::chrono::milliseconds::zero())
+    {
+        auto minutes_left = std::chrono::duration_cast<std::chrono::minutes>(time_left).count();
+
+        attempt_ssh_exec(vm,
+                         fmt::format("wall \"The system is going down for poweroff in {} minute{}, use 'multipass stop "
+                                     "--cancel {}' to cancel the shutdown.\"",
+                                     minutes_left,
+                                     num_plural(minutes_left) ? "s" : "",
+                                     vm.vm_name));
+    }
+    else
+    {
+        attempt_ssh_exec(vm, "wall The system is going down for poweroff now");
     }
 }
 } // namespace
 
-mp::DelayedShutdownTimer::DelayedShutdownTimer(VirtualMachine* virtual_machine, std::optional<SSHSession>&& session,
-                                               const StopMounts& stop_mounts)
-    : virtual_machine{virtual_machine}, ssh_session{std::move(session)}, stop_mounts{stop_mounts}
+mp::DelayedShutdownTimer::DelayedShutdownTimer(VirtualMachine* virtual_machine, const StopMounts& stop_mounts)
+    : virtual_machine{virtual_machine}, stop_mounts{stop_mounts}, time_remaining{0}
 {
 }
 
 mp::DelayedShutdownTimer::~DelayedShutdownTimer()
 {
-    if (shutdown_timer.isActive())
-    {
-        try
+    mp::top_catch_all(virtual_machine->vm_name, [this] {
+        if (shutdown_timer.isActive())
         {
-            if (ssh_session)
-            {
-                // exit_code() is here to make sure the command finishes before continuing in the dtor
-                ssh_session->exec("wall The system shutdown has been cancelled").exit_code();
-            }
-            mpl::log(mpl::Level::info, virtual_machine->vm_name, fmt::format("Cancelling delayed shutdown"));
+            shutdown_timer.stop();
+            mpl::log(mpl::Level::info, virtual_machine->vm_name, "Cancelling delayed shutdown");
             virtual_machine->state = VirtualMachine::State::running;
+            attempt_ssh_exec(*virtual_machine, "wall The system shutdown has been cancelled");
         }
-        catch (const std::exception& e)
-        {
-            mpl::log(mpl::Level::warning, virtual_machine->vm_name,
-                     fmt::format("Unable to cancel delayed shutdown: {}", e.what()));
-            virtual_machine->state = VirtualMachine::State::unknown;
-        }
-    }
+    });
 }
 
-void mp::DelayedShutdownTimer::start(const std::chrono::milliseconds delay)
+void mp::DelayedShutdownTimer::start(std::chrono::milliseconds delay)
```

Comment:
It didn't have to indeed. I saw it as a little drive-by improvement, since I was messing with the interface anyway. But I can put it back if you prefer. 

Because the parameter is passed by value, `const` affects only the variable inside the function, not what is passed to it. I tend to regard this use of `const` as non-canonical and possibly confusing. See also [this](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information).

Reply:
Right, it's not const reference, but my idea is that putting `const` there indicates that it's not something that should be mutable even if it is a local copy to the function since it's being set by the caller.  In that sense, the function should not be able to modify its copy as well.  But as you said, it's a preference.  However, I wouldn't want a modification of the function in the future to modify that value and putting `const` there guarantees that.

Reply:
Yeah, I understand. There is a benefit to having a variable `const`  and that applies to this one too. It is just a pity that in C++ that implementation detail needs to affect the interface... Anyway, I will put it back.

Reply:
Done.